File: design-pattern-jvm/app/src/test/java/design/pattern/java/patterns/structural/flyweight_cache/FlyWeightCacheTest.java
package design.pattern.java.patterns.structural.flyweight_cache;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class FlyWeightCacheTest {
  class ColorShapeMock extends ColorShape {
    private Boolean wasCalled;

    public ColorShapeMock(String color, String shape) {
      super(color, shape);
    }

    @Override
    public Particle move(Particle context, Number x, Number y) {
      this.wasCalled = true;
      return super.move(context, x, y);
    }

    public Boolean hasBeenCalled() {
      return this.wasCalled;
    }
  }

  @Test
  public void shouldReturnUniqueColorShapeWhenIsNotCached() {
    var uniqueState = new ColorShape("red", "circle");
    var particle = new Particle(1, 1, uniqueState);
    Assertions.assertTrue(particle.getColorShape() == uniqueState);
  }

  @Test
  public void shouldReturnCachedColorShapeWhenIsCached() {
    var uniqueState = new ColorShape("red", "circle");
    var cachedState = new ColorShape("red", "circle");
    new Particle(1, 1, uniqueState);
    var particle2 = new Particle(2, 2, cachedState);
    Assertions.assertTrue(particle2.getColorShape() != uniqueState);
  }

  @Test
  public void shouldCallCachedMethodWhenContextMethodIsCalled() {
    final var uniqueState = new ColorShapeMock("red", "circle");
    final var particle = new Particle(3, 3, uniqueState);
    particle.move(3, 3);
    Assertions.assertTrue(uniqueState.hasBeenCalled());
  }

}

File: design-pattern-jvm/app/src/test/java/design/pattern/kotlin/patterns/structural/flyweight_cache/FlyWeightCacheTest.kt
package design.pattern.kotlin.patterns.structural.flyweight_cache

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions

class FlyWeightCacheTest {
    internal inner class ColorShapeMock(color: String, shape: String) : ColorShape(color, shape) {
        private var wasCalled: Boolean? = null
        override fun move(context: Particle, x: Number, y: Number): Particle {
            wasCalled = true
            return super.move(context, x, y)
        }

        fun hasBeenCalled(): Boolean? {
            return wasCalled
        }
    }

    @Test
    fun shouldReturnUniqueColorShapeWhenIsNotCached() {
        val uniqueState = ColorShape("red", "circle")
        val particle = Particle(1, 1, uniqueState)
        Assertions.assertTrue(particle.colorShape === uniqueState)
    }

    @Test
    fun shouldReturnCachedColorShapeWhenIsCached() {
        val uniqueState = ColorShape("red", "circle")
        val cachedState = ColorShape("red", "circle")
        Particle(1, 1, uniqueState)
        val particle2 = Particle(2, 2, cachedState)
        Assertions.assertTrue(particle2.colorShape !== uniqueState)
    }

    @Test
    fun shouldCallCachedMethodWhenContextMethodIsCalled() {
        val uniqueState = ColorShapeMock("red", "circle")
        val particle = Particle(3, 3, uniqueState)
        particle.move(3, 3)
        Assertions.assertTrue(uniqueState.hasBeenCalled()!!)
    }
}